game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/BattleStep.java

```diff
@@ -0,0 +1,99 @@
+package games.strategy.triplea.delegate.battle.steps;
+
+import static games.strategy.triplea.delegate.battle.steps.BattleStep.Request.EXEC;
+
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.IExecutable;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import java.util.List;
+import lombok.AllArgsConstructor;
+
+/**
+ * A step in a battle.
+ *
+ * <ol>
+ *   <ul>
+ *     Each step can have 0 or more names. These names are shown in the {@link
+ *     games.strategy.triplea.ui.BattlePanel}
+ *   </ul>
+ *   <ul>
+ *     Each step can also have an executable. See {@link BattleAtomic} on what an executable is
+ *     comprised of.
+ *   </ul>
+ * </ol>
+ */
+@AllArgsConstructor
+public abstract class BattleStep {
+
+  /**
+   * The current state of the battle
+   *
+   * <p>Note: This is current mutable so the underlying data can change.
+   */
+  protected final BattleState battleState;
+
+  /** Actions that can occur in a battle that require interaction with {@link IDelegateBridge} */
+  protected final BattleActions battleActions;
+
+  /** Indicates when {@link #valid} is being called */
+  public enum Request {
+    // Occurs at the start of the battle round
+    NAME,
+    // Occurs right before the step executes
+    EXEC,
+  }
+
+  public abstract BattleAtomic getExecutable();
+
+  public abstract List<String> getNames();
+
+  /**
+   * Determine if this step should run based on the request
+   *
+   * @param request Indicates when valid is being called
+   * @return true if valid
+   */
+  public abstract boolean valid(Request request);
+
+  /**
+   * Executes the step
+   *
+   * <p>This is called by the BattleAtomic and {@link #valid} has already been checked
+   *
+   * @param stack The current stack of steps
+   * @param bridge DelegateBridge for interacting with the rest of the program
+   */
+  protected abstract void execute(ExecutionStack stack, IDelegateBridge bridge);
+
+  /**
+   * This is used to break up the battle into separate atomic pieces. If there is a network error,
+   * or some other unfortunate event, then we need to keep track of what pieces we have executed,
+   * and what is left to do. Each atomic step is in its own BattleAtomic with the definition of
+   * atomic is that either:
+   *
+   * <ol>
+   *   <li>The code does not call to an IDisplay, IPlayer, or IRandomSource
```

Comment:
I copied this comment from `MustFightBattle#getBattleExecutables` but I think the comment might be outdated.  I think it should reference `IDelegateBridge` instead of `IDisplay, IPlayer, or IRandomSource`.  The only way I see steps can talk to `IDisplay, IPlayer, or IRandomSource` is by going through `IDelegateBridge`.

Reply:
Probably might as well update it so it makes sense. Won't make any more sense to anyone else at this point, and you've about as much context into this as anyone would have after the work you've done here. So if it does not make sense to you now, it probably won't make sense to anyone else (in which case the documentation is not very useful).